### PR TITLE
GPyTorch

### PR DESCRIPTION
**Pull request by miretchin**
It's not finished yet. Need to work out a common framework between the two libraries. But, I made the tests pass for now.

Will have to revise the tests when I complete the API.

### CODE REVIEW

**Comment by miretchin** [comment]
Completed API between libraries. Tests probably no longer pass. Biggest API change -> pulled the loss function outside of model, into the training loop. New module called `marginal_likelihood` houses these losses and works for the `SimpleSupervisedModel` as well.

**Comment by miretchin** [comment]
BTW, I tested the new API with ESOL and I'm able to achieve SOTA. So it maintains the performance identical to the old API.

**Comment by miretchin** [comment]
All the tests are passing! @yuanqing-wang please review when you have a chance.

**Comment by miretchin** [comment]
The only thing we need to investigate is the appending to `Dataset`. It somehow sets a new class init signature once you've appended and every time after that if you call `Dataset()`, it'll give you the updated list, instead of an empty list.

Example:

```python
ds = malt.Dataset()
for idx in range(5):
    point = malt.Molecule(smiles="C", metadata={'y': target[idx].item()})
    ds.append(point)

ds = malt.Dataset()
assert ds.molecules == []
```

**Comment by yuanqing-wang** [comment]
> The only thing we need to investigate is the appending to `Dataset`. It somehow sets a new class init signature once you've appended and every time after that if you call `Dataset()`, it'll give you the updated list, instead of an empty list.
> 
> Example:
> 
> ```python
> ds = malt.Dataset()
> for idx in range(5):
>     point = malt.Molecule(smiles="C", metadata={'y': target[idx].item()})
>     ds.append(point)
> 
> ds = malt.Dataset()
> assert ds.molecules == []
> ```

Yeah this is a classic one: if we have empty list in the __init__ step it's going to be shared among instances. It's a simple fix tho.
